### PR TITLE
feature(ecs-gestalt): Migrate Components to gestalt's Components.

### PR DESCRIPTION
**Pull request by DarkWeird**
### Contains

This is the first step of re-integration gestalt's Entity System (`gestalt-entity-system`) to Terasology by switching to use gestalt's `Component` interface.

This PR does not yet introduce any entity system logic from gestalt, it just switches out Terasology's `Component` marker interface for the one from gestalt.

As gestalt assumes that `Component#copyFrom` is doing a **deep copy**, a potential bug is non-cloning collections and nested types properly at component data copy. Since we're not yet using this functionality we have to keep this pitfall in mind once we're switching to use `gestalt-entity-system` completely.

### How to test

All branches for this migration PR are named `feature/migrate-ecs-to-gestalt`.

```bash
for m in . $(ls -d modules/*/); do                                                                                                                    
	git -C $m remote update -p
	git -C $m switch feature/migrate-ecs-to-gestalt
	git -C $m merge --ff-only @{u}
done
```

* Compile (Ωmega workspace)
  * [x] `gradlew compileJava`
  * [x] `gradlew compileTestJava`
* Run Game
* Start Game
* Run Tests
  * [ ] `gradlew test` 

### Need ToDo

* [ ] Merge https://github.com/MovingBlocks/gestalt/pull/123
* [x] Migrate omega module space (see PRs referencing this PR)

### CODE REVIEW

File: engine/src/main/java/org/terasology/engine/core/bootstrap/EntitySystemSetupUtil.java [hunk 1]
```diff
@@ -165,7 +166,7 @@ private static EventSystem createEventSystem(NetworkSystem networkSystem, PojoEn
 
     private static void registerComponents(ComponentLibrary library, ModuleEnvironment environment) {
         for (Class<? extends Component> componentType : environment.getSubtypesOf(Component.class)) {
-            if (componentType.getAnnotation(DoNotAutoRegister.class) == null) {
+            if (componentType.getAnnotation(DoNotAutoRegister.class) == null && !componentType.isInterface() && !Modifier.isAbstract(componentType.getModifiers())) {
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/bootstrap/EntitySystemSetupUtil.java:169]: why is this change necessary?

File: engine-tests/src/main/java/org/terasology/unittest/stubs/ListOfEnumsComponent.java [hunk 2]
```diff
@@ -2,12 +2,17 @@
...
 import java.util.ArrayList;
 import java.util.List;
 
-public class ListOfEnumsComponent implements Component {
+public class ListOfEnumsComponent implements Component<ListOfEnumsComponent> {
     public List<Side> elements = new ArrayList<>();
+
+    @Override
+    public void copy(ListOfEnumsComponent other) {
+        this.elements = other.elements;
```
Comment (skaldarnar) [inline engine-tests/src/main/java/org/terasology/unittest/stubs/ListOfEnumsComponent.java:16]: Is this a "deep copy"? From the [gesalt wiki](https://github.com/MovingBlocks/gestalt/wiki/Gestalt-Entity-System-Quick-Start)

> Copies should be deep, so that components do not end up sharing references to mutable objects.

**Review by skaldarnar — COMMENTED** [review]
I like that you try to push adoption here now that we have the foundation based on gestalt v7 :+1: 

As mentioned on Discord, I'm looking at the code right now and I'm wondering about the "deep copy" advise given in the gestalt wiki. 
We have lots of components holding collections of elements, and in this first iteration you only do a shallow clone. Is that intentional?

I'm concerned that, if we don't have best-practice implementations in the engine and in Iota, we'll have a hard time fixing those up in the aftermath.

Also, we'll need to update all modules in Omega to not break the workspace. I think we wanted to throw out a couple of modules that are not working anymore, which will at least reduce the scope slightly...

File: engine/src/main/java/org/terasology/engine/core/bootstrap/EntitySystemSetupUtil.java (hunk 1 above)
Comment (DarkWeird) [reply engine/src/main/java/org/terasology/engine/core/bootstrap/EntitySystemSetupUtil.java:169]: Because:
1. `@DoNotAutoRegister` is part of TS, not gestalt.
2.  It fails with `EmptyComponent`. really abstract classes is not functional for ECS.

**Comment by DarkWeird** [comment]
Provide shallow and deep copy for collections.
shallow-copy for collections with immutable copies (String, primitives, enums, EntityRef,Assets,AssetData,Block... and custom immutable types)
deep-copy for several collections, where mutable types is using.

File: engine/build.gradle [hunk 3]
```diff
@@ -103,9 +103,11 @@ dependencies {
...
 
     // Our developed libs
-    api group: 'org.terasology.gestalt', name: 'gestalt-asset-core', version: '7.1.0'
-    api group: 'org.terasology.gestalt', name: 'gestalt-module', version: '7.1.0'
-    api group: 'org.terasology.gestalt', name: 'gestalt-util', version: '7.1.0'
+    api group: 'org.terasology.gestalt', name: 'gestalt-asset-core', version: '7.1.0-SNAPSHOT'
+    api group: 'org.terasology.gestalt', name: 'gestalt-module', version: '7.1.0-SNAPSHOT'
+    api group: 'org.terasology.gestalt', name: 'gestalt-entity-system', version: '7.1.0-SNAPSHOT'
+    api group: 'org.terasology.gestalt', name: 'gestalt-util', version: '7.1.0-SNAPSHOT'
+    api "com.github.zafarkhaja:java-semver:0.10.0"  // ASAP: Remove after https://github.com/MovingBlocks/gestalt/pull/110
```
Comment (skaldarnar) [inline engine/build.gradle:110]: https://github.com/MovingBlocks/gestalt/pull/110 has been merged - do we still need this?

File: engine-tests/src/main/java/org/terasology/unittest/stubs/ListOfObjectComponent.java [hunk 4]
```diff
@@ -3,16 +3,28 @@
...
 
+    @Override
+    public void copy(ListOfObjectComponent other) {
+        this.shortName = other.shortName;
+        this.elements = other.elements.stream().map((subElement -> {
+            SubElement newSub = new SubElement();
+            newSub.id = subElement.id;
+            newSub.type = subElement.type;
+            return newSub;
+        })).collect(Collectors.toList());
```
Comment (skaldarnar) [inline engine-tests/src/main/java/org/terasology/unittest/stubs/ListOfObjectComponent.java:25]: To directly use best-practices in the tests I'd add a copy constructor for `SubElement`.

File: engine/src/jmh/java/org/terasology/benchmark/reflectFactory/GetterSetterComponent.java [hunk 5]
```diff
@@ -2,9 +2,9 @@
 // SPDX-License-Identifier: Apache-2.0
 package org.terasology.benchmark.reflectFactory;
 
-import org.terasology.engine.entitySystem.Component;
+import org.terasology.gestalt.entitysystem.component.Component;
 
-public class GetterSetterComponent implements Component {
+public class GetterSetterComponent implements Component<GetterSetterComponent> {
```
Comment (skaldarnar) [inline engine/src/jmh/java/org/terasology/benchmark/reflectFactory/GetterSetterComponent.java:7]: This is missing a `copy` method.

File: engine/src/main/java/org/terasology/engine/core/bootstrap/EntitySystemSetupUtil.java (hunk 1 above)
Comment (skaldarnar) [reply engine/src/main/java/org/terasology/engine/core/bootstrap/EntitySystemSetupUtil.java:173]: Long line warning, add line break.

File: engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java [hunk 6]
```diff
@@ -184,6 +190,17 @@ private void unregisterPrefabFormats(ModuleAwareAssetTypeManager assetTypeManage
         }
     }
 
+
+    private static void registerComponents(ComponentLibrary library, ModuleEnvironment environment) {
+        for (Class<? extends Component> componentType : environment.getSubtypesOf(Component.class)) {
+            if (componentType.getAnnotation(DoNotAutoRegister.class) == null && !componentType.isInterface() && !Modifier.isAbstract(componentType.getModifiers())) {
+                String componentName = MetadataUtil.getComponentClassName(componentType);
+                Name componentModuleName = verifyNotNull(environment.getModuleProviding(componentType), "Could not find module for %s %s", componentName, componentType);
+                library.register(new ResourceUrn(componentModuleName.toString(), componentName), componentType);
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java:199]: long lines, please add line breaks to match checkstyle conventions.

File: engine/src/main/java/org/terasology/engine/logic/actions/PlaySoundActionComponent.java [hunk 7]
```diff
@@ -28,4 +28,11 @@ public PlaySoundActionComponent() {
     public PlaySoundActionComponent(StaticSound... sounds) {
         this.sounds.addAll(Arrays.asList(sounds));
     }
+
+    @Override
+    public void copy(PlaySoundActionComponent other) {
+        this.sounds = other.sounds;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/logic/actions/PlaySoundActionComponent.java:34]: I think this should create a copy of the list of a deep copy.

File: engine/src/main/java/org/terasology/engine/logic/behavior/BehaviorComponent.java [hunk 8]
```diff
@@ -2,16 +2,22 @@
...
  */
 @API
-public class BehaviorComponent implements Component {
+public class BehaviorComponent implements Component<BehaviorComponent> {
     public BehaviorTree tree;
     public transient Interpreter interpreter;
+
+    @Override
+    public void copy(BehaviorComponent other) {
+        this.tree = other.tree;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/logic/behavior/BehaviorComponent.java:20]: `BehaviorTree` is an asset - do we need to create deep copies of assets, or can we use the same instance?

File: engine/src/main/java/org/terasology/engine/logic/behavior/BehaviorComponent.java [hunk 9]
```diff
@@ -2,16 +2,22 @@
...
+        this.interpreter = other.interpreter;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/logic/behavior/BehaviorComponent.java:21]: `Interpreter` has a copy constructor to create a new instance, but does not do deep copies of any of its members ( e.g., actor or tree). Since it has a copy constructor, should we use it for good measure?

^ @casals any input on this?

File: engine/src/main/java/org/terasology/engine/logic/behavior/CollectiveBehaviorComponent.java [hunk 10]
```diff
@@ -2,16 +2,22 @@
...
 @API
-public class CollectiveBehaviorComponent implements Component {
+public class CollectiveBehaviorComponent implements Component<CollectiveBehaviorComponent> {
     public BehaviorTree tree;
     public transient CollectiveInterpreter collectiveInterpreter;
+
+    @Override
+    public void copy(CollectiveBehaviorComponent other) {
+        this.tree = other.tree;
+        this.collectiveInterpreter = other.collectiveInterpreter;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/logic/behavior/CollectiveBehaviorComponent.java:21]: Same comment as for `BehaviorComponent`.

File: engine/src/main/java/org/terasology/engine/logic/behavior/GroupTagComponent.java [hunk 11]
```diff
@@ -39,4 +40,11 @@
     public BehaviorTree backupBT;
 
     public Interpreter backupRunningState;
+
+    @Override
+    public void copy(GroupTagComponent other) {
+        this.groups = Lists.newArrayList(other.groups);
+        this.backupBT = other.backupBT;
+        this.backupRunningState = other.backupRunningState;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/logic/behavior/GroupTagComponent.java:48]: Same comment as for `BehaviorComponent`.

**Review by skaldarnar — COMMENTED** [review]
Comments on a couple of files. Will push changes for everything where I don't need input on in a minute.

File: engine/src/main/java/org/terasology/engine/logic/behavior/nui/BehaviorNodeComponent.java [hunk 12]
```diff
@@ -27,4 +27,16 @@
...
+
+    @Override
+    public void copy(BehaviorNodeComponent other) {
+        this.action = other.action;
+        this.name = other.name;
+        this.displayName = other.displayName;
+        this.category = other.category;
+        this.shape = other.shape;
+        this.color = other.color;
+        this.textColor = other.textColor;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/logic/behavior/nui/BehaviorNodeComponent.java:39]: use copy constructor of `Color` here

File: engine/src/main/java/org/terasology/engine/logic/characters/CharacterSoundComponent.java [hunk 13]
```diff
@@ -33,4 +32,25 @@
...
+    public void copy(CharacterSoundComponent other) {
+        this.footstepSounds = other.footstepSounds;
+                this.damageSounds = other.damageSounds;
+        this.landingSounds = other.landingSounds;
+                this.jumpSounds = other.jumpSounds;
+        this.deathSounds = other.deathSounds;
+                this.respawnSounds = other.respawnSounds;
+        this.swimSounds = other.swimSounds;
+                this.enterWaterSounds = other.enterWaterSounds;
+        this.leaveWaterSounds = other.leaveWaterSounds;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/logic/characters/CharacterSoundComponent.java:45]: use new array list here? (and fix indentation)

File: engine/src/main/java/org/terasology/engine/logic/common/RetainComponentsComponent.java [hunk 14]
```diff
@@ -15,7 +15,12 @@
...
  * entity still having this component afterwards. If not retained, it is likely to be removed instead.
  */
-public class RetainComponentsComponent implements Component {
+public class RetainComponentsComponent implements Component<RetainComponentsComponent> {
     @Replicate
     public Set<Class<? extends Component>> components = Sets.newHashSet();
+
+    @Override
+    public void copy(RetainComponentsComponent other) {
+        this.components = Sets.newHashSet(other.components); // TODO Investigate, needs to deep-copy or not!
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/logic/common/RetainComponentsComponent.java:24]: probably not? we're on the meta-level here and just reference component classes not component instances.

File: engine/src/main/java/org/terasology/engine/logic/nameTags/NameTagComponent.java [hunk 15]
```diff
@@ -29,4 +29,12 @@
 
     @Replicate
     public float scale = 1f;
+
+    @Override
+    public void copy(NameTagComponent other) {
+        this.yOffset = other.yOffset;
+        this.text = other.text;
+        this.textColor = other.textColor;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/logic/nameTags/NameTagComponent.java:37]: use copy constructor of `Color` here

File: engine/src/main/java/org/terasology/engine/network/ColorComponent.java [hunk 16]
```diff
@@ -3,14 +3,19 @@
...
  */
-public class ColorComponent implements Component {
+public class ColorComponent implements Component<ColorComponent> {
 
     @Replicate
     public Color color;
+
+    @Override
+    public void copy(ColorComponent other) {
+        this.color = other.color;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/network/ColorComponent.java:19]: use copy constructor of `Color` here

File: engine/src/main/java/org/terasology/engine/network/PingStockComponent.java [hunk 17]
```diff
@@ -15,7 +15,7 @@
  * <p>
  * Might be used to stock ping information and display it in future.
  */
-public final class PingStockComponent implements Component {
+public final class PingStockComponent implements Component<PingStockComponent> {
 
     // TODO Map<EntityRef,Long> is not supported for replication (no type handler),
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/network/PingStockComponent.java:20]: We do have https://github.com/MovingBlocks/Terasology/blob/develop/subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandler.java for arbitrary maps by now - can this component be simplified now?

File: engine/src/main/java/org/terasology/engine/particles/components/ParticleEmitterComponent.java [hunk 18]
```diff
@@ -110,4 +110,26 @@
...
+        this.particleSpawnsLeft = other.particleSpawnsLeft;
+        this.destroyEntityWhenDead = other.destroyEntityWhenDead;
+        this.ownerEntity = other.ownerEntity;
+        this.particlePool = other.particlePool;
+        this.generatorFunctionMap.clear();
+        this.generatorFunctionMap.putAll(other.generatorFunctionMap);
+        this.affectorFunctionMap.clear();
+        this.affectorFunctionMap.putAll(other.affectorFunctionMap);
+        this.locationComponent = new LocationComponent();
+        this.locationComponent.copy(other.locationComponent); // TODO check this
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/particles/components/ParticleEmitterComponent.java:131]: This likely works, but I think to remember that @immortius advised us to not have direct references between components. So the "real fix" here would be to change the implementation of the emitter component accordingly.

File: engine/src/main/java/org/terasology/engine/particles/components/ParticleEmitterComponent.java [hunk 19]
```diff
@@ -110,4 +110,26 @@
...
+        this.spawnRateMax = other.spawnRateMax;
+        this.spawnRateMin = other.spawnRateMin;
+        this.enabled = other.enabled;
+        this.lifeTime = other.lifeTime;
+        this.particleSpawnsLeft = other.particleSpawnsLeft;
+        this.destroyEntityWhenDead = other.destroyEntityWhenDead;
+        this.ownerEntity = other.ownerEntity;
+        this.particlePool = other.particlePool;
+        this.generatorFunctionMap.clear();
+        this.generatorFunctionMap.putAll(other.generatorFunctionMap);
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/particles/components/ParticleEmitterComponent.java:127]: Is just putting all values here enough, or do we need to create deep copies here? This again is a direct reference from one component to another, which we should try to avoid.. 

Same holds for `affectorFunctionMap`.

File: engine/src/main/java/org/terasology/engine/rendering/logic/ChunkMeshComponent.java [hunk 20]
```diff
@@ -25,4 +25,11 @@ public ChunkMeshComponent(ChunkMesh mesh, AABBf aabb) {
         this.mesh = mesh;
         this.aabb = aabb;
     }
+
+    @Override
+    public void copy(ChunkMeshComponent other) {
+        this.mesh = other.mesh; // TODO deep or shallow copy?
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/rendering/logic/ChunkMeshComponent.java:31]: we're assuming that we don't need to do deep-copies of assets. if at some point we find proof against this assumption, we'll have to have another look at this

File: engine/src/main/java/org/terasology/engine/rendering/logic/FloatingTextComponent.java [hunk 21]
```diff
@@ -7,10 +7,19 @@
...
     public Color textColor = Color.WHITE;
     public Color textShadowColor = Color.BLACK;
     public float scale = 1f;
     public boolean isOverlay;
+
+    @Override
+    public void copy(FloatingTextComponent other) {
+        this.text = other.text;
+        this.textColor = other.textColor;
+        this.textShadowColor = other.textShadowColor;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/rendering/logic/FloatingTextComponent.java:21]: use copy constructor of `Color` here

File: engine/src/main/java/org/terasology/engine/rendering/logic/MeshComponent.java [hunk 22]
```diff
@@ -33,4 +33,13 @@
...
     public Color color = Color.WHITE;
 
+    @Override
+    public void copy(MeshComponent other) {
+        this.mesh = other.mesh;
+        this.material = other.material;
+        this.translucent = other.translucent;
+        this.selfLuminance = other.selfLuminance;
+        this.hideFromOwner = other.hideFromOwner;
+        this.color = other.color;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/rendering/logic/MeshComponent.java:43]: use copy constructor of `Color` here

File: engine/src/main/java/org/terasology/engine/rendering/logic/RegionOutlineComponent.java [hunk 23]
```diff
@@ -8,8 +8,15 @@
...
+public class RegionOutlineComponent implements VisualComponent<RegionOutlineComponent> {
     public Vector3i corner1;
     public Vector3i corner2;
     public Color color = new Color(Color.white);
+
+    @Override
+    public void copy(RegionOutlineComponent other) {
+        this.corner1 = new Vector3i(other.corner1);
+        this.corner2 = new Vector3i(other.corner2);
+        this.color = other.color;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/rendering/logic/RegionOutlineComponent.java:20]: use copy constructor of `Color` here

**Review by skaldarnar — COMMENTED** [review]
Some general assumptions we put on this while reviewing:

- we don't have to create copies of immutable types (this applies to enums, but also to something like `EntityRef`)
- we don't have to create copies of assets (is that true?) 
- we should use a copy constructor wherever possible
- we trust existing copy constructors to do the right job

_(this list might be extended or adjusted in the future)_

File: engine/src/main/java/org/terasology/engine/rendering/logic/SkeletalMeshComponent.java [hunk 24]
```diff
@@ -56,4 +57,21 @@
...
+        this.loop = other.loop;
+        this.animationPool = Lists.newArrayList(other.animationPool);
+        this.animationRate = other.animationRate;
+        this.heightOffset = other.heightOffset;
+        this.boneEntities = Maps.newHashMap(other.boneEntities);
+        this.rootBone = other.rootBone;
+        this.animationTime = other.animationTime;
+        this.scale = new Vector3f(other.scale);
+        this.translate = new Vector3f(other.translate);
+        this.color = other.color;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/rendering/logic/SkeletalMeshComponent.java:75]: use copy constructor of `Color` here

File: engine/src/main/java/org/terasology/engine/telemetry/GamePlayStatsComponent.java [hunk 25]
```diff
@@ -27,4 +28,13 @@
...
     @Replicate
     public Map<String, Integer> creatureKilled = new HashMap<>();
+
+    @Override
+    public void copy(GamePlayStatsComponent other) {
+        this.blockDestroyedMap = Maps.newHashMap(other.blockDestroyedMap);
+        this.blockPlacedMap = Maps.newHashMap(other.blockPlacedMap);
+        this.distanceTraveled = other.distanceTraveled;
+        this.playTimeMinute = other.playTimeMinute;
+        this.creatureKilled = other.creatureKilled;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/telemetry/GamePlayStatsComponent.java:38]: use `newHashMap` here

File: engine/src/main/java/org/terasology/engine/world/block/BlockComponent.java [hunk 26]
```diff
@@ -54,4 +54,10 @@ public Vector3i getPosition(Vector3i dest) {
         dest.set(position);
         return dest;
     }
+
+    @Override
+    public void copy(BlockComponent other) {
+        this.position = new Vector3i(other.position);
+        this.block = other.block;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/world/block/BlockComponent.java:61]: Do we need to deep-copy `Block`?

File: engine/src/main/java/org/terasology/engine/world/block/items/BlockItemComponent.java [hunk 27]
```diff
@@ -21,4 +21,9 @@ public BlockItemComponent() {
     public BlockItemComponent(BlockFamily blockFamily) {
         this.blockFamily = blockFamily;
     }
+
+    @Override
+    public void copy(BlockItemComponent other) {
+        this.blockFamily = other.blockFamily;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/world/block/items/BlockItemComponent.java:27]: Do we need to deep-copy `BlockFamily`?

File: engine/src/main/java/org/terasology/engine/world/block/regions/ActAsBlockComponent.java [hunk 28]
```diff
@@ -3,14 +3,20 @@
...
  *
  */
-public final class ActAsBlockComponent implements Component {
+public final class ActAsBlockComponent implements Component<ActAsBlockComponent> {
     public BlockFamily block;
     public boolean dropBlocksInRegion;
+
+    @Override
+    public void copy(ActAsBlockComponent other) {
+        this.block = other.block;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/world/block/regions/ActAsBlockComponent.java:19]: Do we need to deep-copy `BlockFamily`?
And why is that member called "block" if it's a block family :roll_eyes:

File: engine/src/main/java/org/terasology/engine/world/block/typeEntity/BlockTypeComponent.java [hunk 29]
```diff
@@ -2,11 +2,16 @@
...
+import org.terasology.gestalt.entitysystem.component.Component;
 
-public class BlockTypeComponent implements Component {
+public class BlockTypeComponent implements Component<BlockTypeComponent> {
     @Replicate
     public Block block;
+
+    @Override
+    public void copy(BlockTypeComponent other) {
+        this.block = other.block;
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/world/block/typeEntity/BlockTypeComponent.java:15]: Do we need to deep-copy `Block`?

File: engine/src/main/java/org/terasology/engine/world/block/items/BlockItemComponent.java (hunk 27 above)
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/world/block/items/BlockItemComponent.java:27]: I don't think you can modify a BlockFamily. its treated as a singleton.

File: engine/src/main/java/org/terasology/engine/world/block/BlockComponent.java (hunk 26 above)
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/world/block/BlockComponent.java:61]: block is also a immutable object.

Comment (skaldarnar) [reply engine/src/main/java/org/terasology/engine/world/block/BlockComponent.java:61]: https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/engine/world/block/Block.java has quite a few setters mutating internal state for an _immutable object_ ... 🤔

**Comment by DarkWeird** [comment]
> Some general assumptions we put on this while reviewing:
> 
> - we don't have to create copies of immutable types (this applies to enums, but also to something like `EntityRef`)
> - we don't have to create copies of assets (is that true?) 
> - we should use a copy constructor wherever possible
> - we trust existing copy constructors to do the right job
> 
> _(this list might be extended or adjusted in the future)_

@skaldarnar @jdrueckert 

4. I don't trust copy constructors :)
     They can be non-valid for technical use. They can do not full copy, then part of functional with non-copy field can not works.

3. I prefer to use special 'copy' method (see answer for 4)

1,2. Yeah. You can don't create copy for immutable objects or shared/pooled objects:
EntityRef, Block, Asset,Prefabs, BlockFamily(iirc)

Shared/pooled objects can be replaced by string/uri (prefabs, asset, block, blockfamily)

File: engine-tests/src/test/java/org/terasology/engine/persistence/typeHandling/reflection/TestModuleEnvironmentSandbox.java [hunk 30]
```diff
@@ -53,7 +53,7 @@ public void findSubtypeOfGlobalClass() {
 
     @Test
     public void findSubtypeOfEngineClass() {
-        assertNotEmpty(sandbox.findSubTypeOf("engine:BlockComponent", Component.class));
```
Comment (skaldarnar) [inline engine-tests/src/test/java/org/terasology/engine/persistence/typeHandling/reflection/TestModuleEnvironmentSandbox.java:56]: Is it a problem that we can no longer resolve components by their simple identifiers when trying to resolve subtypes via the module sandbox?

**Comment by jdrueckert** [comment]
[Migrate to gestalt components](https://trello.com/c/aumKFb9w/150-migrate-to-gestalt-components)

**Comment by skaldarnar** [comment]
I've renamed `copy` >>> `copyFrom` for the engine now and bumped the dependency for gestalt to `7.2.0-SNAPSHOT`.

This depends on https://github.com/MovingBlocks/gestalt/pull/123

I've also applied the rename to all modules in my Ωmega workspace.

File: engine/build.gradle (hunk 3 above)
Comment (skaldarnar) [reply engine/build.gradle:120]: ^ pinging @keturn @DarkWeird

**Review by jdrueckert — APPROVED** [review]
